### PR TITLE
Add promu support to build binary

### PR DESCRIPTION
**Pull request by CharlesJUDITH**
- Modify Makefile for promu
- Add VERSION file
- Add promu.yml

### CODE REVIEW

**Comment by CharlesJUDITH** [comment]
Regarding the travis build you'll need something like:
```yaml
sudo: required

services:
- docker

language: go

go:
- 1.7.x
- 1.8.x
- master

matrix:
  allow_failures:
  - go: master

script:
- make style test

before_deploy:
- make promu
- promu crossbuild
- promu crossbuild tarballs
- promu checksum .tarballs

deploy:
  provider: releases
  api_key:
    secure: $GITHUB_OAUTH_TOKEN
  file_glob: true
  file: .tarballs/*
  skip_cleanup: true
  on:
    tags: true
```
There is a useful documentation here: https://docs.travis-ci.com/user/build-stages/deploy-github-releases/

File: Makefile [hunk 1]
```diff
@@ -1,9 +1,17 @@
 NAME    := unsee
 VERSION := $(shell git describe --tags --always --dirty='-dev')
+GO      := GO15VENDOREXPERIMENT=1 go
```
Comment (prymitive) [inline Makefile]: Go >=1.8 is required to compile unsee, there's no point in setting `GO15VENDOREXPERIMENT`

File: Makefile [hunk 2]
```diff
@@ -23,14 +31,14 @@ endif
 
 .build/deps-build-go.ok:
 	@mkdir -p .build
-	go get -u github.com/golang/dep/cmd/dep
-	go get -u github.com/jteeuwen/go-bindata/...
-	go get -u github.com/elazarl/go-bindata-assetfs/...
+	$(GO) get -u github.com/golang/dep/cmd/dep
```
Comment (prymitive) [inline Makefile]: There no need to use `$(GO)` here because there no need for any extra envs passed to the go binary

File: VERSION [hunk 3]
```diff
@@ -0,0 +1 @@
+0.0.8
```
Comment (prymitive) [inline VERSION]: Currently version comes from the git tag, there's no workflow to maintain this file and I don't think there's any need for it

File: Makefile [hunk 4]
```diff
@@ -52,13 +60,30 @@ bindata_assetfs.go: .build/deps-build-go.ok .build/artifacts-bindata_assetfs.$(G
...
 
+build: promu
+	@echo ">> building binaries"
+	@$(PROMU) build --prefix $(PREFIX)
+
+tarball: promu
+	@echo ">> building release tarball"
+	@$(PROMU) tarball --prefix $(PREFIX) $(BIN_DIR)
+
+docker:
```
Comment (prymitive) [inline Makefile]: There's already a target for building docker images

File: Makefile [hunk 5]
```diff
@@ -52,13 +60,30 @@ bindata_assetfs.go: .build/deps-build-go.ok .build/artifacts-bindata_assetfs.$(G
...
-	go build -ldflags "-X main.version=$(VERSION)"
+	$(GO) build -ldflags "-X main.version=$(VERSION)"
 
 .build/vendor.ok: .build/deps-build-go.ok Gopkg.lock Gopkg.toml
 	dep ensure
 	dep prune
 	touch $@
 
+build: promu
+	@echo ">> building binaries"
```
Comment (prymitive) [inline Makefile]: There's no `@echo` usage for any other target, no need to make any new target special in that regard

File: Makefile [hunk 6]
```diff
@@ -1,9 +1,17 @@
 NAME    := unsee
 VERSION := $(shell git describe --tags --always --dirty='-dev')
+GO      := GO15VENDOREXPERIMENT=1 go
+PROMU   := $(GOPATH)/bin/promu
+pkgs     = $(shell $(GO) list ./... | grep -v -E '/vendor/')
+
+PREFIX                  ?= $(shell pwd)
+BIN_DIR                 ?= $(shell pwd)
+DOCKER_IMAGE_NAME       ?= unsee
```
Comment (prymitive) [inline Makefile]: There'a already a target for building docker images, there's no need for an extra variables for it

File: Makefile [hunk 7]
```diff
@@ -1,9 +1,17 @@
 NAME    := unsee
 VERSION := $(shell git describe --tags --always --dirty='-dev')
+GO      := GO15VENDOREXPERIMENT=1 go
+PROMU   := $(GOPATH)/bin/promu
+pkgs     = $(shell $(GO) list ./... | grep -v -E '/vendor/')
```
Comment (prymitive) [inline Makefile]: where is this used?

File: .promu.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
+repository:
+    path: github.com/cloudflare/unsee
+build:
+    binaries:
+        - name: unsee
+          path: .
+    flags: -a -tags netgo
+    ldflags: |
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.Version={{.Version}}
```
Comment (prymitive) [inline .promu.yml]: I don't understand `{{repoPath}}/vendor/github.com/prometheus/common/version.` part here, why is it needed?

Comment (prymitive) [reply .promu.yml]: As you can see on https://github.com/cloudflare/unsee/blob/master/Makefile#L55 there's only one linker flag passed currently

File: .promu.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
+repository:
+    path: github.com/cloudflare/unsee
+build:
+    binaries:
+        - name: unsee
+          path: .
+    flags: -a -tags netgo
```
Comment (prymitive) [inline .promu.yml]: what does `-a -tags netgo` do?

File: .promu.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    ldflags: |
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.Version={{.Version}}
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.Revision={{.Revision}}
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.Branch={{.Branch}}
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.BuildUser={{user}}@{{host}}
+        -X {{repoPath}}/vendor/github.com/prometheus/common/version.BuildDate={{date "20060102-15:04:05"}}
+
+tarball:
+    files:
+        - docs/example.yaml
```
Comment (prymitive) [inline .promu.yml]: shouldn't it be entire `docs` directory and the `README.md` file?

**Comment by prymitive** [comment]
promu seems very Prometheus oriented, @terinjokes  uses a more simple tool which doesn't require so much customization and can be fully baked into .travis.yaml which, given that the goal is to upload those binaries only from travis, seems like a really good idea 
https://github.com/terinjokes/bakelite/blob/master/.travis.yml#L20

File: Makefile (hunk 7 above)
Comment (prymitive) [reply Makefile]: Those are rather intrusive changes to the current workflow and they are not in scope for the ticket this aims to solve

File: Makefile (hunk 2 above)
Comment (prymitive) [reply Makefile]: But those are practices that Prometheus developers setup for the projects they maintain, there is no need to make intrusive workflow changes here just to comply with other projects practices

File: Makefile (hunk 5 above)
Comment (prymitive) [reply Makefile]: Again, those are not practices established here, nor is there any reason to comply here

File: VERSION (hunk 3 above)
Comment (prymitive) [reply VERSION]: I don't like this pattern and I don't want to manage any version tracking files, git tags are the source of truth for that right now and I didn't hit any issues with that. `promu` seems very intrusive tool that creates a set of requirements and constrains that I don't feel entirely happy with. I think I prefer the `bakealite` approach

**Comment by prymitive** [comment]
I gave bakelite a try and it doesn't require any Makefile changes, all commands for it target Travis and it's contained within `.travis.yml` which I feel more comfortable with - #212
Thanks for raising this but I prefer bakelite solution

**Comment by CharlesJUDITH** [comment]
No problem. Thanks for your feedback.

**Pull request closed without merging**